### ISSUE TITLE
Popup previews for Wikitionary hang

### ISSUE THREAD

**Issue by ZebraFromPluto** [gnosygnu/xowa#511]
This issue only affects Wikitionary. All the other wikis work fine. I have full downloads of the HTML, Wikitext source files, and images files for all the main wikis. en.wikipedia.org (2018-07), simple.wikipedia.org (2017-05), and (2017-04) for the rest.

Quite simply, cursoring over links usually takes extraordinary amounts of time to build the popup previews. Sometimes the popups appear in a reasonable time, but it never takes much playing around to get the long build times. It can take about 15 seconds to over two minutes to appear. During this time time it will flash many different "load page for ..." messages in the bottom corner, often hanging on certain messages for extended durations. Also, the CPU and memory usage for javaw.exe go through the roof during these times. Before long, javaw.exe goes from using around 200 MB to over 2 GB. Playing in the other wikis doesn't do this, just Wikitionary.

Usually, with enough playing around, I will eventually get a window pop up as shown here:
![Wiktionary_problem](https://user-images.githubusercontent.com/26853004/61068388-e5c99400-a3d7-11e9-83a3-9899138498db.png)
I have "Eclipse Mars.1 Release (4.5.1)" installed on my computer, and it appears to be interfering.

Although I've done work on code that was Java, I'm really not savvy with Java in general. Just so you understand things from my perspective, I've been a coder since the early 80's, but have never delved into modern web languages and the like. I pretty much just work in C-based languages.

Do you experience this behaviour as well, or is it something on my end, and if so, do you have any idea how to address the problem? Moving the Wikitext files out of the folder will allow everything to work okay, displaying "retrieving data" instead of the popup, but obviously a proper solution would be ideal.

**Comment by gnosygnu** [comment]
Hi,

Thanks again for the details. The short answer is you can work around this by reducing the Popup functionality for en.wiktionary only.

Specifically:
* Go to ```home/wiki/Special:XowaCfg?grp=xowa.addon.popups```
* Scroll down to ```Template keeplist```
* Replace the contents with the following: 
```
en.wikipedia.org|formatnum;age;age_in_days;age_in_years_and_days*;nts;number_table_sorting*;as_of;oldstyledatedy;gregorian_serial_date;currentminute;currentsecond;dmca;spaced_ndash;trim;month*;convert*;worldpop*;ipa*;lang*;nowrap*;h:*;mvar;math;vgy;audio;iso_639_name;transl;translate;linktext;zh;nihongo*;japanese_name;ko-hhrm;mp|
en.wiktionary.org|some_random_text_to_ignore_all_templates|
``` 
* Restart XOWA
* Go to any en.wiktonary.org pages and hover -> Popups will be much quicker

Other details below. Let me know if you still run into issues.

Thanks.

---

> Quite simply, cursoring over links usually takes extraordinary amounts of time to build the popup previews. 
>  It can take about 15 seconds to over two minutes to appear. 

Yeah, I'm not surprised. 
* en.wiktionary.org makes use of a lot of templates. 
* XOWA Popups "re-compile" the HTML for a page when hovering over a link for the page
* Depending on the page, this could mean a noticeable delay. en.wiktionary.org/wiki/steam took about 30 seconds for me, but I'm on an SSD

> do you have any idea how to address the problem?

The process above basically disables parsing for all templates in en.wiktionary.org. Specifically, it'll only parse a template called ```some_random_text_to_ignore_all_templates``` which doesn't exist. Since no templates will be parsed, the popups will just compile the wikitext which will be much faster. The downside is that popups will be missing information, but it should hopefully be good enough.

> Although I've done work on code that was Java, I'm really not savvy with Java in general. Just so you understand things from my perspective, I've been a coder since the early 80's, but have never delved into modern web languages and the like. I pretty much just work in C-based languages.

Okay, good to know. I've only dabbled in C, but do have a healthy respect for those who work at such a low-level. No garbage collection, no built-in exception handling, and no real class-based semantics were my recollections of my struggles with it.

**Comment by ZebraFromPluto** [comment]
That works beautifully! Thank you. Would it perhaps be best to default like that? (BTW, I spelled "Wiktionary" incorrectly in my original post. Oops.)

Although the etymology doesn't come through, I think the fast method may contain more useful information, at least for the word "cave" I checked here:
![before_and_after](https://user-images.githubusercontent.com/26853004/61128718-b6bb2d00-a480-11e9-9a64-d154f6efd09a.png)

As for my C background, I spent several years programming in just assembly language, so C seemed very easy by comparison. Most modern languages are largely built on C, so I can manage with them when the need arises. Dealing with pointers, error handling, and structuring code without the language doing it for you are second nature to me so doesn't bother me at all. But certainly the more modern languages are a better choice overall today.
